File: solo-leveling-ai-service/src/main/java/com/sleepkqq/sololeveling/ai/kafka/GenerateTasksConsumer.java
package com.sleepkqq.sololeveling.ai.kafka;

import static com.sleepkqq.sololeveling.avro.constants.KafkaGroupIds.AI_GROUP_ID;
import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.GENERATE_TASKS_TOPIC;

import com.sleepkqq.sololeveling.ai.service.ChatService;
import com.sleepkqq.sololeveling.avro.task.GenerateTasksEvent;
import com.sleepkqq.sololeveling.avro.task.SaveTasksEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class GenerateTasksConsumer {

  private final ChatService chatService;
  private final SaveTasksProducer saveTasksProducer;

  @KafkaListener(topics = GENERATE_TASKS_TOPIC, groupId = AI_GROUP_ID)
  public void listen(GenerateTasksEvent event) {
    log.info(">> Start tasks generation | transactionId={}", event.getTransactionId());
    var generatedTasks = StreamEx.of(event.getInputs())
        .parallel()
        .map(chatService::generateTask)
        .toList();

    log.info("<< Tasks successfully generated | transactionId={}", event.getTransactionId());
    saveTasksProducer.send(new SaveTasksEvent(
        event.getTransactionId(),
        event.getPlayerId(),
        generatedTasks
    ));
  }
}

File: solo-leveling-ai-service/src/main/kotlin/com/sleepkqq/sololeveling/ai/kafka/GenerateTasksConsumer.kt
package com.sleepkqq.sololeveling.ai.kafka

import com.sleepkqq.sololeveling.ai.service.ChatService
import com.sleepkqq.sololeveling.avro.config.consumer.AbstractKafkaConsumer
import com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics
import com.sleepkqq.sololeveling.avro.idempotency.IdempotencyService
import com.sleepkqq.sololeveling.avro.task.GenerateTasksEvent
import com.sleepkqq.sololeveling.avro.task.SaveTasksEvent
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.annotation.RetryableTopic
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class GenerateTasksConsumer(
	private val chatService: ChatService,
	private val saveTasksProducer: SaveTasksProducer,
	idempotencyService: IdempotencyService
) : AbstractKafkaConsumer<GenerateTasksEvent>(
	idempotencyService = idempotencyService,
	log = LoggerFactory.getLogger(GenerateTasksConsumer::class.java)
) {

	@Transactional
	@RetryableTopic
	@KafkaListener(
		topics = [KafkaTaskTopics.GENERATE_TASKS_TOPIC],
		groupId = $$"${spring.kafka.avro.group-id}"
	)
	fun listen(event: GenerateTasksEvent) {
		consumeWithIdempotency(event)
	}

	override fun getTxId(event: GenerateTasksEvent): String = event.txId

	override fun processEvent(event: GenerateTasksEvent) {
		log.info(">> Start tasks generation | txId={}", event.txId)
		val generatedTasks = event.inputs.map { chatService.generateTask(it) }

		log.info("<< Tasks successfully generated | txId={}", event.txId)
		saveTasksProducer.send(
			SaveTasksEvent(
				event.txId,
				event.playerId,
				generatedTasks
			)
		)
	}
}
